moveit_setup_assistant/src/tools/moveit_config_data.cpp

```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
   return true;  // file created successfully
 }
 
+// ******************************************************************************************
+// Output controllers config files
+// ******************************************************************************************
+bool MoveItConfigData::outputControllersYAML(const std::string& file_path)
+{
+  // Cache the joints' names.
+  std::vector<std::vector<std::string>> groups;
+  std::vector<std::string> groupJoints;
+
+  // Loop through groups
+  for (std::vector<srdf::Model::Group>::iterator group_it = srdf_->groups_.begin(); group_it != srdf_->groups_.end();
+       ++group_it)
+  {
+    // Get list of associated joints
+    const robot_model::JointModelGroup* joint_model_group = getRobotModel()->getJointModelGroup(group_it->name_);
+    const std::vector<const robot_model::JointModel*>& joint_models = joint_model_group->getActiveJointModels();
+    // Iterate through the joints
+    for (const robot_model::JointModel* joint : joint_models)
+    {
+      if (joint->isPassive() || joint->getMimic() != NULL || joint->getType() == robot_model::JointModel::FIXED)
+        continue;
+      else
+      {
+        groupJoints.push_back(joint->getName());
+      }
+    }
+    groups.push_back(groupJoints);
+  }
+
+  YAML::Emitter emitter;
+  emitter << YAML::BeginMap;
+
+  emitter << YAML::Comment("Settings for ros_control control loop");
+  emitter << YAML::Key << "generic_hw_control_loop" << YAML::BeginMap;
+  emitter << YAML::Key << "loop_hz";
+  emitter << YAML::Value << "300";
+  emitter << YAML::Key << "cycle_time_error_threshold";
+  emitter << YAML::Value << "0.01";
+  emitter << YAML::EndMap;
+  emitter << YAML::EndMap;
+
+  emitter << YAML::Comment("Settings for ros_control hardware interface");
+  emitter << YAML::BeginMap;
+  emitter << YAML::Key << "hardware_interface";
+  emitter << YAML::Value << YAML::BeginMap;
+
+  emitter << YAML::Key << "groups";
+  emitter << YAML::Value << YAML::BeginSeq;
+
+  // Loop through groups
+  for (int i = 0; i < srdf_->groups_.size(); ++i)
+  {
+    emitter << YAML::BeginMap;
+    emitter << YAML::Key << "name";
+    emitter << YAML::Value << srdf_->groups_[i].name_ + "_controller";
+    emitter << YAML::Key << "joints";
+    emitter << YAML::Value << YAML::BeginSeq;
+
+    // Iterate through the joints
+    for (int j = 0; j < groups[i].size(); j++)
+    {
+      emitter << groups[i][j];
+    }
+    emitter << YAML::EndSeq;
+    emitter << YAML::EndMap;
+  }
+  emitter << YAML::EndSeq;
+
+  emitter << YAML::Key << "sim_control_mode";
+  emitter << YAML::Value << "1";
+  emitter << YAML::Comment("0: position, 1: velocity");
+
+  emitter << YAML::EndMap;
+
+  emitter << YAML::Newline;
+  emitter << YAML::Comment("Publish all joint states");
+  emitter << YAML::Newline << YAML::Comment("Creates the /joint_states topic necessary in ROS");
+  emitter << YAML::EndMap;
+
+  emitter << YAML::BeginMap;
```

Comment:
what if in order to track the begin and end of maps, brackets were used to indent these sections? e.g.

```
emitter << YAML::BeginMap;
{
    emitter << YAML::Key << "joint_state_controller";
    emitter << YAML::Value << YAML::BeginMap;
    ...
}
emitter << YAML::EndMap;
```